src/test/function_test/test_split.cpp

```diff
@@ -0,0 +1,356 @@
+            std::string expected_value;
+            auto ret = pg_client->get(hash_key, sort_key, expected_value);
+            ASSERT_EQ(ret, PERR_OK);
+            ASSERT_EQ(expected[hash_key][sort_key], expected_value);
+        }
+    }
+
+    void hash_scan_during_split(int32_t count)
+    {
+        if (count % 3 != 0) {
+            return;
+        }
+
+        pegasus_client::pegasus_scanner *scanner = nullptr;
+        pegasus_client::scan_options options;
+        auto ret = pg_client->get_scanner(
+            dataset_hashkey_prefix + std::to_string(count), "", "", options, scanner);
+        ASSERT_TRUE((ret == PERR_OK || ret == PERR_TIMEOUT));
+        if (ret == PERR_OK) {
+            std::string hash_key;
+            std::string sort_key;
+            std::string expected_value;
+            while (scanner->next(hash_key, sort_key, expected_value) == 0) {
+                ASSERT_EQ(expected[hash_key][sort_key], expected_value);
+            }
+        }
+        delete scanner;
+    }
+
+    void full_scan_after_split()
+    {
+        std::cout << "Start full scan after partition split......" << std::endl;
+        std::vector<pegasus_client::pegasus_scanner *> scanners;
+        pegasus_client::scan_options options;
+        auto ret = pg_client->get_unordered_scanners(10000, options, scanners);
+        ASSERT_EQ(ret, PERR_OK);
+        int32_t count;
+        for (auto i = 0; i < scanners.size(); i++) {
+            std::string hash_key;
+            std::string sort_key;
+            std::string expected_value;
+            pegasus_client::internal_info info;
+            pegasus_client::pegasus_scanner *scanner = scanners[i];
+            while (scanner->next(hash_key, sort_key, expected_value, &info) == 0) {
+                ASSERT_EQ(expected[hash_key][sort_key], expected_value);
+                count++;
+            }
+        }
+        ASSERT_EQ(count, dataset_count);
+        for (auto scanner : scanners) {
+            delete scanner;
+        }
+    }
+
+public:
+    std::shared_ptr<replication_ddl_client> ddl_client;
+    pegasus_client *pg_client;
+    std::unordered_map<std::string, std::unordered_map<std::string, std::string>> expected;
+    const int32_t partition_count = 4;
+    const int32_t dataset_count = 1000;
+    const std::string dataset_hashkey_prefix = "hashkey";
+    const std::string dataset_sortkey_prefix = "sortkey";
+    const std::string splitting_hashkey_prefix = "keyh_";
+    const std::string splitting_sortkey_prefix = "keys_";
+    const std::string data_value = "vaule";
+    int32_t count_during_split = 0;
+};
+
+TEST_F(partition_split_test, split_with_write)
+{
+    const std::string table_name = "split_table_1";
+    SetUp(table_name);
+    start_partition_split(table_name);
+    // write data during partition split
+    do {
+        write_data_during_split();
+        std::this_thread::sleep_for(std::chrono::seconds(1));
+    } while (!is_split_finished(table_name));
```

Comment:
How to guarantee the loop starts before the split process ends

Reply:
I don't figure out what is your comment mean, could you please explain more about it?